### PR TITLE
Pfacctests/packetfabric cloud router connection aws

### PR DESCRIPTION
**Pull request by pontovinte**
## Description

ACCTest for packetfabric_cloud_router_connection_aws

### CODE REVIEW

**Comment by rjouhann** [comment]
Hello @pontovinte @sousafeliciano 

I went ahead and updated the PR with the following changes (see [commit](https://github.com/PacketFabric/terraform-provider-packetfabric/pull/424/commits/deb82dd3366abd08c5413709341a458927b49110)):

1. `resource_aws_cloud_router_connection_test.go`:
- Replaced `crConnAwsResult := testutil.RHclCloudRouterConnectioAws()` `to crConnAwsResult := testutil.RHclCloudRouterConnectionAws()`
- Added `testutil.PF_ACCOUNT_ID_KEY,`
- Removed `testutil.PF_CRC_SPEED_KEY,`
- Added `ExternalProviders: testAccExternalProviders,`
- Added `resource.TestCheckResourceAttr(crConnAwsResult.ResourceName, "pop", crConnAwsResult.Pop),`
- Modified the test so the import state test works (adding `ImportStateIdFunc` and setting `cloudRouterCircuitId` and `cloudRouterConnectionCircuitId`.

2. `test_envar_keys.go`:
- Replaced `const PF_CRC_AWS_ACCOUNT_ID_KEY = "PF_CRC_AWS_ACCOUNT_ID"` with `const PF_CRC_AWS_ACCOUNT_ID_KEY = "PF_AWS_ACCOUNT_ID"`

3. `test_hcls.go`:
- Removed duplicated function `func RHclCloudRouterConnectioAws() RHclCloudRouterConnectionAWSResult {}`
- Added `Speed` and `AccountUuid` to `type RHclCloudRouterConnectionAwsResult struct {}`
- Added `AccountUuid:  os.Getenv(PF_ACCOUNT_ID_KEY),` and `Speed:        CloudRouterConnAwsSpeed,` to `RHclCloudRouterConnectionAws()`

However, I am still getting an error:
```
5780b9be90e8# TF_ACC=1 go test -v -run=TestAccCloudRouterConnectionAws
=== RUN   TestAccCloudRouterConnectionAwsRequiredFields
pop === PAUSE TestAccCloudRouterConnectionAwsRequiredFields
=== CONT  TestAccCloudRouterConnectionAwsRequiredFields
    resource_aws_cloud_router_connection_test.go:16: Step 1/2 error: Error running apply: exit status 1
        
        Error: Status: 400, body: {"message": "pop: This field is required.", "request_id": "d6353146-77d3-42d2-ae86-78bcda91dfe8"}
        
          with packetfabric_cloud_router_connection_aws.pf_3a8eae90_07a2_4bf5_8fc8_fb4ee73bf42f,
          on terraform_plugin_test.tf line 15, in resource "packetfabric_cloud_router_connection_aws" "pf_3a8eae90_07a2_4bf5_8fc8_fb4ee73bf42f":
          15: resource "packetfabric_cloud_router_connection_aws" "pf_3a8eae90_07a2_4bf5_8fc8_fb4ee73bf42f" {
        
--- FAIL: TestAccCloudRouterConnectionAwsRequiredFields (17.37s)
```

I think the problem is  `_findAvailableCloudPopZoneAndMedia()`:
1. I tried adding `fmt.Printf("pop: %v", pop)` after we call the function and it doesn't return anything
2. When I manually set the pop in the test code, the ACC test works! ✅  (`const pop = "LAB4"`)
```
5780b9be90e8# TF_ACC=1 go test -v -run=TestAccCloudRouterConnectionAws
=== RUN   TestAccCloudRouterConnectionAwsRequiredFields
[DEBUG] pop: LAB4
=== PAUSE TestAccCloudRouterConnectionAwsRequiredFields
=== CONT  TestAccCloudRouterConnectionAwsRequiredFields
--- PASS: TestAccCloudRouterConnectionAwsRequiredFields (134.41s)
PASS
ok  	github.com/PacketFabric/terraform-provider-packetfabric/internal/provider	134.427s
```

Also, an unrelated question, why do we need those constants?
```
const PF_CRC_SPEED_KEY = "PF_CRC_SPEED"
const PF_CRC_POP1_KEY = "PF_CRC_POP1"
const PF_CRC_ZONE1_KEY = "PF_CRC_ZONE1"
```

File: internal/testutil/test_hcls.go [hunk 1]
```diff
@@ -218,6 +231,7 @@ func RHclCloudRouterConnectionAws() RHclCloudRouterConnectionAwsResult {
 	}
 
 	pop, _, _ := portDetails._findAvailableCloudPopZoneAndMedia()
```
Comment (rjouhann) [inline internal/testutil/test_hcls.go:233]: @pontovinte @sousafeliciano the logic to get the pop isn't working. see [comment](https://github.com/PacketFabric/terraform-provider-packetfabric/pull/424#issuecomment-1513798696).

Comment (pontovinte) [reply internal/testutil/test_hcls.go:233]: @rjouhann, actually I think the problem is this:
```
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-W-ER4-2 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-W-LR4-18 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-W-CWDM4-2 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-W-LR4-16 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-W-ER-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-W-ER DWDM-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-W-ZR-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-W-ZR DWDM-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-W-LR-39 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-V-ER4-2 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-V-LR4-18 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-V-CWDM4-2 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-V-LR4-16 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-V-ER DWDM-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-V-ZR-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-V-ZR DWDM-1 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-V-LR-48 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-T-LX-78 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-T-ZX-10 ] 
[PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-T-EX-8 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-A-LR4-1 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-A-LR-11 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-D-EX-2 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-D-LX-9 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 1Gbps-D-ZX-2 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-D-ER4-2 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-D-LR4-18 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-D-CWDM4-2 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 100Gbps-D-LR4-16 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-D-ER-1 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-D-ER DWDM-1 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-D-ZR-1 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-D-ZR DWDM-1 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-D-LR-36 ] 
[PORT | DESIRED POP: LAB1 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 10Gbps-B-LR-14 ] 
```
There's no `POP` available for `50Mbps`.

Comment (rjouhann) [reply internal/testutil/test_hcls.go:233]: > @rjouhann, actually I think the problem is this:
> 
> ```
> [PORT | DESIRED POP: LAB4 | DESIRED SPEED: 50Mbps | AVAILABLE SPEED-ZONE-MEDIA-COUNT: 40Gbps-W-ER4-2 ] 
> 
> There's no `POP` available for `50Mbps` being sent from API.

@pontovinte 

I think there is a misunderstanding about how you select the pop for Cloud Router Connections. The speed is hardcoded to 50Mbps.

Selecting the pop for cloud router connection is different than selecting the pop for a port. I think this is where the confusion is.

- **For Cloud Router Connection:**

Example with AWS:
GET /v2/locations/cloud?has_cloud_router=true&cloud_provider=aws&cloud_connection_type=hosted
```
 {
        "pop": "LAB4",
        "region": "US",
        "market": "GOG",
        "market_description": "us-central1",
        "zones": [
            "W"
        ],
        "vendor": "Google Compute",
        "site": "PF LAB 4",
        "site_code": "PF-LAB4",
        "type": "primary",
        "status": "Active",
        "latitude": "25.000000",
        "longitude": "71.000000",
        "timezone": null,
        "notes": null,
        "pcode": null,
        "lead_time": "instant",
        "single_armed": false,
        "address1": "gce-us-central1",
        "address2": null,
        "city": "Google",
        "state": "GO",
        "postal": "4444",
        "country": "US",
        "cloud_provider": "aws",
        "cloud_connection_details": {
            "region": "us-west-1",
            "hosted_type": "hosted-connection",
            "region_description": "US West (N. California)"
        },
        "network_provider": "PacketFabric",
        "time_created": "2019-04-08T22:58:02.000000+0000",
        "enni_supported": true
    },
```
As you can see, there are no speed, no media to select from there, you only need to use of the pop (here example LAB4).

- **For Port:** (also see comment [here](https://github.com/PacketFabric/terraform-provider-packetfabric/pull/356#issuecomment-1486893561))
First, get the list of pop available: GET /v2/locations
Example:
```
{
        "pop": "LAB1",
        "region": "US",
        "market": "LAB",
        "market_description": "Dallas LAB1",
        "vendor": "PacketFabric",
        "site": "PacketFabric LAB1",
        "site_code": "PF-LAB1",
        "type": "primary",
        "status": "Active",
        "latitude": "32.795581",
        "longitude": "-96.783443",
        "timezone": null,
        "notes": null,
        "pcode": 3725900,
        "lead_time": "instant",
        "single_armed": false,
        "address1": "1649 W Frankford Rd",
        "address2": "",
        "city": "Carrollton",
        "state": "TX",
        "postal": "75007",
        "country": "US",
        "network_provider": "PacketFabric",
        "time_created": null,
        "enni_supported": true
    },
```

Then get the availability: GET /v2/locations/LAB1/port-availability
```
[
    {
        "speed": "100Gbps",
        "media": "LR4",
        "zone": "A",
        "partial": false,
        "enni": false,
        "count": 1
    },
    {
        "speed": "10Gbps",
        "media": "LR",
        "zone": "A",
        "partial": false,
        "enni": false,
        "count": 11
    },
    {
        "speed": "1Gbps",
        "media": "EX",
        "zone": "D",
        "partial": false,
        "enni": true,
        "count": 2
    },
    {
        "speed": "1Gbps",
        "media": "LX",
        "zone": "D",
        "partial": false,
        "enni": true,
        "count": 9
    },
    {
        "speed": "1Gbps",
        "media": "ZX",
        "zone": "D",
        "partial": false,
        "enni": true,
        "count": 2
    },
]
```

From there, ONLY select the port with speed = 1Gbps, then use the media/zone info to create the port.

Comment (rjouhann) [reply internal/testutil/test_hcls.go:233]: @pontovinte @elXaptation I have fixed the logic to select a cloud pop for a cloud router connection.

I create 2 new functions: `_findAvailableCloudPopZone()` and `FetchCloudPopsAndZones()`. The new logic works in both `dev` and `prod`.

Output:
```
cf9143733951# TF_ACC=1 go test -v -run=TestAccCloudRouterConnectionAws
=== RUN   TestAccCloudRouterConnectionAwsRequiredFields
2023/05/08 22:16:32 Starting to search for available Cloud PoP and zone...
2023/05/08 22:16:32 Available PoPs with Zones: map[AMS1:[A B] ATL1:[C D E F] CDG2:[A B] CHI1:[E F] CHI11:[C D] DAL1:[C D] DEN1:[E F] FRA1:[A B] LAS1:[A B] LAX1:[C F] LON1:[A B] MCI1:[A B] MIA1:[C D] NYC1:[C D] PDX2:[A B] PHX5:[A B] SEA3:[A B] SFO6:[A B] SYD1:[A B] TOR2:[A B] WDC1:[E F] WDC2:[C E F]]
2023/05/08 22:16:32 Checking PoP: SEA3
2023/05/08 22:16:32 Found available PoP: SEA3, Zone: A
=== PAUSE TestAccCloudRouterConnectionAwsRequiredFields
=== CONT  TestAccCloudRouterConnectionAwsRequiredFields
--- PASS: TestAccCloudRouterConnectionAwsRequiredFields (187.04s)
PASS
ok  	github.com/PacketFabric/terraform-provider-packetfabric/internal/provider	187.064s

```

**Review by rjouhann — APPROVED** [review]

**Comment by rjouhann** [comment]
@medzin could you give this PR another review? Tests passed in prod and dev.

**prod**
```
cf9143733951# TF_ACC=1 go test -v -run=TestAccCloudRouterConnectionAws
=== RUN   TestAccCloudRouterConnectionAwsRequiredFields
2023/05/08 22:16:32 Starting to search for available Cloud PoP and zone...
2023/05/08 22:16:32 Available PoPs with Zones: map[AMS1:[A B] ATL1:[C D E F] CDG2:[A B] CHI1:[E F] CHI11:[C D] DAL1:[C D] DEN1:[E F] FRA1:[A B] LAS1:[A B] LAX1:[C F] LON1:[A B] MCI1:[A B] MIA1:[C D] NYC1:[C D] PDX2:[A B] PHX5:[A B] SEA3:[A B] SFO6:[A B] SYD1:[A B] TOR2:[A B] WDC1:[E F] WDC2:[C E F]]
2023/05/08 22:16:32 Checking PoP: SEA3
2023/05/08 22:16:32 Found available PoP: SEA3, Zone: A
=== PAUSE TestAccCloudRouterConnectionAwsRequiredFields
=== CONT  TestAccCloudRouterConnectionAwsRequiredFields
--- PASS: TestAccCloudRouterConnectionAwsRequiredFields (187.04s)
PASS
ok  	github.com/PacketFabric/terraform-provider-packetfabric/internal/provider	187.064s
```

**dev**
```
=== RUN   TestAccCloudRouterConnectionAwsRequiredFields
2023/05/08 22:27:53 Starting to search for available Cloud PoP and zone...
2023/05/08 22:27:53 Available PoPs with Zones: map[DEV1:[A B] LAB1:[A] LAB4:[W]]
2023/05/08 22:27:53 Checking PoP: LAB1
2023/05/08 22:27:53 Found available PoP: LAB1, Zone: A
=== PAUSE TestAccCloudRouterConnectionAwsRequiredFields
=== CONT  TestAccCloudRouterConnectionAwsRequiredFields
--- PASS: TestAccCloudRouterConnectionAwsRequiredFields (184.11s)
PASS
ok  	github.com/PacketFabric/terraform-provider-packetfabric/internal/provider	184.135s
```

**Review by rjouhann — APPROVED** [review]

**Review by rjouhann — APPROVED** [review]

**Pull request merged**